src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int StackProbingThreshold = 100;
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _recursionDepth;
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
+        {
+            if (++_recursionDepth >= StackProbingThreshold)
+            {
+                try
+                {
+                    RuntimeHelpers.EnsureSufficientExecutionStack();
+                }
+                catch (InsufficientExecutionStackException) when (_executionStackCount < MaxExecutionStackCount)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public void Exit()
+        {
+            _recursionDepth--;
+        }
+
+        public void RunOnEmptyStack(Action<object> action, object state)
+        {
```

Comment:
Unfortunately, `TaskFactory.StartNew` does not have an overload that's generic in the `state` parameter. Adding @stephentoub to chime in; I've seen the same pattern in `TaskExtensions`.